### DISCUSSION TITLE
2nd factor once a week for paper and each time for live?

### DISCUSSION THREAD

**Discussion by akosmaroy** [IbcAlpha/IBC#271] [General]
Hi,

I'm new to IBC.

Reading the documentation gives me the impression that one would need to use 2nd factor authentication once a week when using the TWS gateway & IBC. But my experience is that when logging in to paper trading indeed it's not asking for the 2nd factor each time, but when starting the TWS Gateway via IBC for live / real trading, it does ask for the 2nd factor each time.

Am I missing something?

Ákos

**Comment by rlktradewright** [comment]
Yes, you are definitely missing something.

First of all, paper trading accounts do not use 2nd factor authentication at all.

Secondly, for the live session, if it’s behaving unexpectedly as you say, please submit your IBC logfile so I can see what’s going on. Send the whole thing – don’t try to edit out stuff you think is irrelevant. Attach the file to your reply, rather than quoting it inline.

**Comment by akosmaroy** [comment]
here you go - thank you for looking into it

[ibc-3.20.0_GATEWAY-1030_Monday.txt](https://github.com/user-attachments/files/17621705/ibc-3.20.0_GATEWAY-1030_Monday.txt)

**Comment by rlktradewright** [comment]
You have this in your `config.ini`:

```
AutoLogoffTime=10:10 PM
AutoRestartTime=
```

This means that Gateway will shut down completely every day at 10:10 pm and you then have to restart it., which means loading it from scratch. Starting Gateway from scratch always needs a full 2FA login.

Instead you need to use auto-restart, like this:

```
AutoLogoffTime=
AutoRestartTime=10:10 PM
```

Then, when the AutoRestartTime is reached, Gateway creates a file called `autorestart` which contains the cryptographically sealed credentials needed to re-establish the login session, and then shuts down. The IBC script now creates a new instance of Gateway which detects the `autorestart` file and uses it to re-create the original login session without any user involvement.

**Comment by akosmaroy** [comment]
I see. It seems then I misunderstood the process.

As I only need the gateway during trading hours on NYSE, I was hoping to start it from cron each day before trading starts and shut it down after trading ends, each day. And I was hoping this would work with only one 2FA per week. It seems I was mistaken, and have to keep it running, with or without auto restart?

**Comment by rlktradewright** [comment]
You don't have to keep it running if you don't use auto-restart. So you can certainly use cron to start it before the start of the trading session.

Frankly, the need to do 2FA every day if not auo-restarting doesn't seem too onerous to me. You just have to make sure your phone is nearby and switched on. - it could hardly be easier. You don't even have to get out of bed!

Having said that, it would be technically feasible as an option to configure it with auto-restart and then have the IBC script NOT create a new instance, just exit. Then when you restart the next time (using cron or whatever) the `autorestart` file would still exist and the new Gateway would attempt to use it to re-establish the session. However, I believe Gateway will object if the credentials in the `autorestart` have expired, and I don't know how quickly that happens. My feeling is that the expiry period should be short (ie a matter of a few seconds) to avoid the possibility of the `autorestart` file being somehow stolen, but I know that isn't the case. I might spend some time trying to figure out how long it can be. For example, if an `autorestart` file generated at close of session one day is still valid at start of session the next day, then what you want could be a possibility... (but it would be vulnerable to IBKR shortening the expiry period).

**Comment by akosmaroy** [comment]
Thanks for the explanation
